launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,500 @@
 #![no_std]
 
 elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use elrond_wasm::elrond_codec::TopEncode;
+
+mod setup;
+
+mod launch_stage;
+use launch_stage::*;
+
+mod ongoing_operation;
+use ongoing_operation::*;
+
+mod random;
+use random::Random;
+
+mod ticket_status;
+use ticket_status::TicketStatus;
+
+const VEC_MAPPER_START_INDEX: usize = 1;
+const FIRST_GENERATION: u8 = 1;
 
 #[elrond_wasm::derive::contract]
-pub trait Launchpad {
-    #[init]
-    fn init(&self) {}
+pub trait Launchpad: setup::SetupModule + ongoing_operation::OngoingOperationModule {
+    // endpoints - owner-only
+
+    #[only_owner]
+    #[endpoint(claimTicketPayment)]
+    fn claim_ticket_payment(&self) -> SCResult<()> {
+        let ticket_payment_token = self.ticket_payment_token().get();
+        let sc_balance = self.blockchain().get_sc_balance(&ticket_payment_token, 0);
+        let owner = self.blockchain().get_caller();
+
+        if sc_balance > 0 {
+            self.send()
+                .direct(&owner, &ticket_payment_token, 0, &sc_balance, &[]);
+        }
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[endpoint(forceClaimPeriodStart)]
+    fn force_claim_period_start(&self) -> SCResult<()> {
+        let total_winning_tickets = self.nr_winning_tickets().get();
+        self.total_confirmed_tickets().set(&total_winning_tickets);
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[endpoint(addTickets)]
+    fn add_tickets(
+        &self,
+        #[var_args] address_number_pairs: VarArgs<MultiArg2<Address, usize>>,
+    ) -> SCResult<()> {
+        self.require_no_ongoing_operation()?;
+        self.require_stage(LaunchStage::AddTickets)?;
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let winner_selection_start_epoch = self.winner_selection_start_epoch().get();
+        require!(
+            current_epoch < winner_selection_start_epoch,
+            "Cannot add more tickets, winner selection has started"
+        );
+
+        for multi_arg in address_number_pairs.into_vec() {
+            let (buyer, nr_tickets) = multi_arg.into_tuple();
+
+            self.try_create_tickets(&buyer, nr_tickets)?;
+        }
+
+        Ok(())
+    }
+
+    // endpoints
+
+    #[endpoint(selectWinners)]
+    fn select_winners(&self) -> SCResult<OperationCompletionStatus> {
+        self.require_stage(LaunchStage::SelectWinners)?;
+
+        let last_ticket_position = self.shuffled_tickets().len();
+        let nr_winning_tickets = self.nr_winning_tickets().get();
+
+        // dummy, will be overwritten in the Load part,
+        // but the Rust compiler complains of possibly uninitialized variables otherwise
+        let mut rng = Random::from_hash(self.crypto(), H256::zero(), 0);
+        let mut ticket_position = 0;
+
+        let run_result = self.run_while_it_has_gas(|gas_op| match gas_op {
+            GasOp::Load(op) => match op {
+                OngoingOperationType::None => {
```

Comment:
What happens when you have selectwinners - confirm and select winners again. Is the ongoing operation type nonce ? do we have to reintialize it again ?
I think it would be better on line 89 to have something like : getOrCreateRandom - which verifies if there was random seed saved under a predefined key or not. If it was not, it creates and saves.
I think that would make it safer.

Reply:
Add get or create random seed selection before the run_while_it_has_gas